### PR TITLE
Add YAML adapter and Python API entrypoints

### LINKED ISSUE #6: YAML adapter + Python API entrypoints
Load YAML into Resource objects.
Provide programmatic plan/apply functions.

### PR DESCRIPTION
**Pull request by jonas-meyer**
## Summary

- Add `dss_provisioner.config` module with YAML loading (`ruamel.yaml`), type discrimination, and convenience `plan`/`apply`/`plan_and_apply` functions
- Add `ProviderConfig` (pydantic-settings, env var support) and `Config` models for declarative configuration
- Add `default_registry()` factory wiring all built-in resource types + handlers
- Refactor resource models: rename `dataset_type`/`recipe_type` → `type`, add `yaml_alias` ClassVar for YAML type discrimination
- Refactor recipe handlers: composition over inheritance, extract shared code payload helpers, make `RecipeHandler` generic
- Add `PlanContext` for cross-resource validation lookups (bridges desired resources + state)
- Add `Resource.reference_names()` for automatic dependency inference (recipes auto-depend on their inputs/outputs)
- Refactor engine `plan()`: extract `_resolve_deps`, `_classify_change`, `_plan_deletes` helpers; no longer mutates Resource objects
- Add `ApplyError` for Terraform-style partial failure handling with state recovery

## Test plan

- [x] YAML loader: type discrimination for all dataset/recipe types, error cases, edge cases (234 lines)
- [x] Config API: engine wiring, env var resolution, code file integration (105 lines)
- [x] Registry: all 8 resource types registered with correct handlers
- [x] Partial failure recovery: `ApplyError` carries partial result, re-run picks up from saved state
- [x] All existing tests updated for model field renames
- [x] 229 tests pass, 92% coverage

Closes #6

### CODE REVIEW

File: src/dss_provisioner/config/__init__.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    auth = (
+        ApiKeyAuth(api_key=SecretStr(config.provider.api_key)) if config.provider.api_key else None
+    )
+    provider = DSSProvider(host=config.provider.host, auth=auth)
+    return DSSEngine(
+        provider=provider,
+        project_key=config.provider.project,
+        state_path=config.state_path,
+        registry=default_registry(),
+    )
```
[the remark this answers is not in the document]
Comment (jonas-meyer) [reply src/dss_provisioner/config/__init__.py]: Added early validation — `_engine_from_config` now raises `ConfigError` with a clear message if `host` or `api_key` are missing.
